File: src/motion_spec/dashboard/compare.py
# SPDX-License-Identifier: MPL-2.0

"""Two generations side by side: what the model says, what happened, what the values did.

Three axes, always together and always in that order. A duration that moved with no model
change is a flaky run or a real nondeterminism, and telling those two apart is the whole
value -- a behaviour diff with its model diff missing cannot be read at all.

Nothing here computes a statistic of its own. Coordination comes from the runtime-graph
views, signals from the frame-log reports, and this module only aligns and subtracts. No
delta is ranked or coloured: a longer state may be the fix, and the dashboard cannot know.
"""

from __future__ import annotations

from pathlib import Path

import rdflib
from rdflib.compare import graph_diff

from motion_spec.dashboard.analysis import run_reports
from motion_spec.dashboard.catalog import rdf_name
from motion_spec.dashboard.graph import MODEL_GRAPH, load_model_graph, model_manifest
from motion_spec.dashboard.queries import DECLARED_ACTIVITIES, gates
from motion_spec.dashboard.queries import compare as compare_runs
from motion_spec.introspection.archive import ArchiveError
from motion_spec.introspection.replay import resolve_archive

# What 018 restricts its statistics to, said in the payload so a reader knows the table is
# not every recorded quantity.
WATCHED = "signals a constraint or monitor judges"


# -- the model axis ------------------------------------------------------------------------
#
# Over the graph, never over the .robmot text: a reordered file or a renamed generation must
# not read as a change. rdflib canonicalises both sides before subtracting, so a blank node
# that moved but says the same thing does not either.


def _design_graph(generation_dir: Path) -> rdflib.Graph:
    """A generation's design graph with its generation provenance taken out.

    That provenance is the whole of what a regenerate moves -- the activity IRIs, the
    generatedAtTime stamps, the file:// location of every output -- and it is a closed island
    no design triple points into, so dropping it costs the diff nothing else.
    """
    # Local: one namespace constant, with the DSL's parser generator behind the import.
    from motion_spec_dsl.gens import DSLPROV

    manifest = model_manifest(generation_dir)
    if manifest is None:
        return rdflib.Graph()
    dataset = rdflib.Dataset(default_union=True)
    model = load_model_graph(manifest, dataset, dataset.graph(MODEL_GRAPH))
    design = rdflib.Graph()
    for triple in model:
        if not any(str(term).startswith(DSLPROV) for term in (triple[0], triple[2])):
            design.add(triple)
    return design


def _literal(term):
    """A literal as something JSON can carry; anything else as the term it is."""
    if isinstance(term, rdflib.Literal):
        value = term.toPython()
        return value if isinstance(value, (bool, int, float, str)) else str(term)
    return str(term)


def _by_subject(graph: rdflib.Graph) -> dict:
    index: dict = {}
    for subject, predicate, obj in graph:
        index.setdefault(subject, {}).setdefault(predicate, []).append(obj)
    return index


def _term(subject, changed: rdflib.Graph) -> dict:
    """A term that appeared or vanished whole, read off the diff that holds all its triples.

    Canonicalisation gives a blank node a hash for a label, which names nothing a reader can
    look up, so an anonymous term answers with what it is instead of with that hash.
    """
    types = sorted(rdf_name(kind) for kind in changed.objects(subject, rdflib.RDF.type))
    if isinstance(subject, rdflib.BNode):
        return {"iri": "", "name": " ".join(types) or "anonymous", "types": types}
    return {"iri": str(subject), "name": rdf_name(subject), "types": types}


def _coordination_elements(graph: rdflib.Graph) -> set[str]:
    """What the design says can be occupied at all -- the two sides' shared identity."""
    return {str(row[0]) for row in graph.query(DECLARED_ACTIVITIES)}


def model_diff(left_dir: Path, right_dir: Path) -> dict:
    """What changed in the design graph, classified by what the reader can act on.

    A value that moved is the common case and the most legible row: the term, the property,
    the number before and the number after. A term that appeared or vanished is next. What
    is left is structure -- an operand repointed, a controller set changed -- and it is
    reported as what went and what came rather than pretended to be a value.
    """
    left, right = _design_graph(left_dir), _design_graph(right_dir)
    _both, only_left, only_right = graph_diff(left, right)
    gone, came = _by_subject(only_left), _by_subject(only_right)
    left_terms, right_terms = set(left.subjects()), set(right.subjects())
    values, structure = [], []
    for subject in sorted(gone.keys() | came.keys(), key=str):
        if subject not in right_terms or subject not in left_terms:
            continue  # the whole term went or came; it is reported as that, not as an edit
        before, after = gone.get(subject, {}), came.get(subject, {})
        for predicate in sorted(before.keys() | after.keys(), key=str):
            was, now = before.get(predicate, []), after.get(predicate, [])
            row = {
                "iri": str(subject),
                "name": rdf_name(subject),
                "predicate": str(predicate),
                "property": rdf_name(predicate),
            }
            if len(was) == len(now) == 1 and all(
                isinstance(term, rdflib.Literal) for term in (*was, *now)
            ):
                values.append({**row, "left": _literal(was[0]), "right": _literal(now[0])})
            else:
                structure.append(
                    {
                        **row,
                        "removed": [_literal(term) for term in was],
                        "added": [_literal(term) for term in now],
                    }
                )
    return {
        "same_model": _coordination_elements(left) == _coordination_elements(right),
        "values": sorted(values, key=lambda row: (row["name"], row["property"])),
        "removed": sorted(
            (_term(subject, only_left) for subject in gone if subject not in right_terms),
            key=lambda row: row["name"],
        ),
        "added": sorted(
            (_term(subject, only_right) for subject in came if subject not in left_terms),
            key=lambda row: row["name"],
        ),
        "structure": sorted(structure, key=lambda row: (row["name"], row["property"])),
    }


# -- the coordination axis -----------------------------------------------------------------


def _header(run_dir: Path):
    """The run's own contract header, or nothing when it recorded no frame log.

    A run can be started with logging off; it still occupied states and still has a runtime
    graph, so the coordination axis answers and the signal axis says it cannot.
    """
    try:
        return resolve_archive(Path(run_dir))[3].header
    except (ArchiveError, OSError):
        return None


def _reached_end(spans: list[dict], header) -> bool | None:
    """Whether the run ever occupied the element its contract names as the end."""
    if header is None or not 0 <= header.end_state < len(header.fsm_states):
        return None
    return any(span["element"] == header.fsm_states[header.end_state].iri for span in spans)


def _platform(header) -> str:
    return f"{header.platform_name or 'unnamed'}{' (simulated)' if header.simulated else ''}"


def _get(side: dict | None, key: str):
    return None if side is None else side[key]


def _delta(before, after):
    return None if before is None or after is None else round(after - before, 3)


def _gate_pair(left: dict | None, right: dict | None) -> dict:
    named = left or right
    return {
        "monitor": named["monitor"],
        "name": named["monitor_name"],
        "event": named["event_name"],
        "left_waited_s": _get(left, "waited_s"),
        "right_waited_s": _get(right, "waited_s"),
        "delta_s": _delta(_get(left, "waited_s"), _get(right, "waited_s")),
        "left_rearms": _get(left, "rearm_count"),
        "right_rearms": _get(right, "rearm_count"),
        "left_fired": None if left is None else left["fired_step"] is not None,
        "right_fired": None if right is None else right["fired_step"] is not None,
    }


def _gate_rows(left_run: Path, right_run: Path) -> list[dict]:
    """Each gate's arming on both sides, aligned by monitor, a re-arming its own row."""
    pending: dict[str, list[dict]] = {}
    for row in gates(right_run)["gates"]:
        pending.setdefault(row["monitor"], []).append(row)
    rows = []
    for row in gates(left_run)["gates"]:
        queue = pending.get(row["monitor"]) or []
        rows.append(_gate_pair(row, queue.pop(0) if queue else None))
    rows += [_gate_pair(None, row) for queue in pending.values() for row in queue]
    return rows


def coordination_diff(left_run: Path, right_run: Path) -> dict:
    """What happened, aligned: occupancy durations, control-flow order, gate waits and re-arms.

    Order first. A state visited that was not visited before means the two runs took
    different paths, and durations along different paths are not comparable at all -- so the
    payload says the order differs before it says anything about a duration.
    """
    aligned = compare_runs(left_run, right_run)
    spans = {side: aligned[side]["spans"] for side in ("left", "right")}
    order = {side: [span["element"] for span in rows] for side, rows in spans.items()}
    return {
        "order_differs": order["left"] != order["right"],
        "left_order": [span["name"] for span in spans["left"]],
        "right_order": [span["name"] for span in spans["right"]],
        "left_source": aligned["left"]["runtime_source"],
        "right_source": aligned["right"]["runtime_source"],
        # Whether each run ever occupied what its own contract names as the end.
        "left_reached_end": _reached_end(spans["left"], _header(left_run)),
        "right_reached_end": _reached_end(spans["right"], _header(right_run)),
        "activities": aligned["activities"],
        "gates": _gate_rows(left_run, right_run),
    }


# -- the signal axis -----------------------------------------------------------------------
#
# Aligned on an occupancy, not on a state: a state entered twice is two rows, because a
# re-entry appearing or disappearing is the change worth seeing.


def _occupancies(reports: dict) -> dict[int, tuple[str, int]]:
    """Span index -> (activity, which entry of it), so the two sides align occupancy by
    occupancy rather than by summing everything one state ever did."""
    seen: dict[str, int] = {}
    keys = {}
    for span in reports["states"]:
        seen[span["state_id"]] = nth = seen.get(span["state_id"], -1) + 1
        keys[span["state_index"]] = (span["state_id"], nth)
    return keys


def _paired(left: dict, right: dict, kind: str, identity) -> dict:
    """Both sides' rows of one report kind, keyed by occupancy and by what the row is about."""
    paired: dict = {}
    for side, reports in (("left", left), ("right", right)):
        keys = _occupancies(reports)
        for row in reports[kind]:
            key = keys.get(row["state_index"])
            if key is not None:
                paired.setdefault((key, identity(row)), {})[side] = row
    return paired


def _sides(pair: dict, *fields) -> dict:
    """One row's numbers on both sides, with the deltas of whatever both sides recorded.

    Sign and magnitude, never a verdict: which direction is desirable is the reader's to say.
    """
    left, right = pair.get("left"), pair.get("right")
    out = {}
    for field in fields:
        before, after = _get(left, field), _get(right, field)
        out[f"left_{field}"] = before
        out[f"right_{field}"] = after
        out[f"delta_{field}"] = None if before is None or after is None else after - before
    return out


def _named(pair: dict, field: str):
    """What a row is of, which either side answers the same way."""
    return (pair.get("left") or pair["right"])[field]


def _during(paired: dict, key) -> list:
    """One occupancy's rows of a report kind, ordered by what each row is about."""
    rows = [(what, pair) for (occupancy, what), pair in paired.items() if occupancy == key]
    return sorted(rows, key=lambda row: str(row[0]))


def _saturation(left: dict, right: dict) -> list[dict]:
    """Torque clipping per joint, aligned as 018 reports it -- per run, not per occupancy.

    A clipped joint carries the states it was clipped in rather than a span, so this row sits
    outside the per-activity tables instead of being re-derived into one.
    """
    paired: dict = {}
    for side, reports in (("left", left), ("right", right)):
        for row in reports["saturation"]:
            paired.setdefault((row["solver"], row["joint"]), {})[side] = row
    return [
        {
            "solver": solver,
            "joint": joint,
            "left_states": _get(pair.get("left"), "states") or [],
            "right_states": _get(pair.get("right"), "states") or [],
            **_sides(pair, "ticks", "requested_nm", "limit_nm"),
        }
        for (solver, joint), pair in sorted(paired.items())
    ]


def signal_diff(left_run: Path, right_run: Path) -> dict:
    """Per occupancy, what the values a gate judges did on each side, and by how much it moved.

    Restricted to the signals 018 already restricts itself to; the payload says so, because
    a table that quietly left something out reads as a table of everything.
    """
    left, right = run_reports(left_run), run_reports(right_run)
    watched = _paired(left, right, "signals", lambda row: row["signal"])
    contact = _paired(left, right, "contact", lambda _row: "")
    modes = _paired(left, right, "oscillation", lambda row: row["solver"])
    order = list(dict.fromkeys([*_occupancies(left).values(), *_occupancies(right).values()]))
    activities = []
    for key in order:
        entry = {
            "name": key[0],
            "entry": key[1],
            "signals": [
                {
                    "signal": signal,
                    "gate": _named(pair, "gate"),
                    "band": _named(pair, "band"),
                    **_sides(pair, "peak", "pk_pk", "rms"),
                }
                for signal, pair in _during(watched, key)
            ],
            "contact": [
                {
                    "gate": _named(pair, "gate"),
                    "band": _named(pair, "band"),
                    **_sides(pair, "peak_n", "rebound", "ripple", "settle_s"),
                }
                for _one, pair in _during(contact, key)
            ],
            "modes": [
                {"solver": solver, **_sides(pair, "mode_hz", "mean_amplitude", "agreeing")}
                for solver, pair in _during(modes, key)
            ],
        }
        if any(entry[kind] for kind in ("signals", "contact", "modes")):
            activities.append(entry)
    return {
        "restricted_to": WATCHED,
        "activities": activities,
        "saturation": _saturation(left, right),
    }


# -- what the three axes are worth -----------------------------------------------------------


def compare_generations(
    left: Path, right: Path, left_run: Path | None = None, right_run: Path | None = None
) -> dict:
    """Two generations along all three axes, and a plain statement of what does not line up.

    Runs are optional: two generations with no run between them still have a model diff, and
    saying "these two models differ by one gain" is worth more than refusing. Where a
    comparison is unsound it is said at the top and what aligns is still shown -- refusing is
    unhelpful, and pretending is worse.
    """
    model = model_diff(left, right)
    notes = []
    if not model["same_model"]:
        notes.append("the two sides declare different coordination elements; only what aligns")
    comparable = {
        "same_model": model["same_model"],
        "same_schema_hash": None,
        "same_platform": None,
        "both_completed": None,
        "notes": notes,
    }
    payload = {
        "comparable": comparable,
        "model": model,
        "coordination": {"order_differs": False, "activities": [], "gates": []},
        "signals": {"restricted_to": WATCHED, "activities": [], "saturation": []},
    }
    if left_run is None or right_run is None:
        notes.append("no run picked on both sides; the model diff is all there is to show")
        return payload
    coordination = coordination_diff(left_run, right_run)
    payload["coordination"] = coordination
    # Said before any duration: durations along a different path are not comparable at all.
    if coordination["order_differs"]:
        notes.insert(0, "the two runs took different paths; durations are not comparable")
    for side in ("left", "right"):
        if coordination[f"{side}_source"] != "archive":
            notes.append(
                f"the {side} run has no archived runtime graph; its waits and re-arms are "
                "unavailable, not zero"
            )
    comparable["both_completed"] = bool(
        coordination["left_reached_end"] and coordination["right_reached_end"]
    )
    for side, run in (("left", left_run), ("right", right_run)):
        if coordination[f"{side}_reached_end"] is False:
            notes.append(f"{Path(run).name} did not reach its end state")
    before, after = _header(left_run), _header(right_run)
    if before is None or after is None:
        notes.append("a side recorded no frame log; nothing can be said about its signals")
        return payload
    comparable["same_schema_hash"] = before.schema_hash == after.schema_hash
    if not comparable["same_schema_hash"]:
        notes.append("the two contracts differ; a slot index may not mean the same thing")
    comparable["same_platform"] = _platform(before) == _platform(after)
    if not comparable["same_platform"]:
        notes.append(
            f"{_platform(before)} against {_platform(after)}; "
            "signal figures across platforms are not like for like"
        )
    payload["signals"] = signal_diff(left_run, right_run)
    return payload

File: src/motion_spec/dashboard/frontend/compare.js
// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: 2026 SECORO AG (secoro.uni-bremen.de)

/**
 * Did this change alter behaviour? Two generations, three axes, in one order.
 *
 * Model first, then coordination, then signals -- a duration that moved with no model change
 * is a flaky run or a real nondeterminism, and that is the distinction the page exists for.
 * Nothing here is coloured or ranked: which direction is desirable is the reader's to say.
 */

import { $, api, state } from "./core.js";
import { section, table } from "./views.js";

export const COMPARE_MARKUP =
  '<div class="views"><div class="compare-bar">'
  + '<span class="compare-left"></span>'
  + '<select class="compare-gen"></select><select class="compare-run"></select>'
  + '</div><div class="compare-slot"></div></div>';

const dash = "—";

// Enough digits to see a gain move, few enough to scan a column of them.
function num(value) {
  if (value === null || value === undefined) return dash;
  if (typeof value !== "number") return String(value);
  if (value === 0) return "0";
  const size = Math.abs(value);
  return size >= 1e4 || size < 1e-3 ? value.toExponential(2) : String(Number(value.toPrecision(4)));
}

const signed = (value) =>
  value === null || value === undefined ? dash : `${value >= 0 ? "+" : "−"}${num(Math.abs(value))}`;

const seconds = (value) => (value === null || value === undefined ? dash : `${value.toFixed(2)} s`);

// The tail of an IRI is what a table cell can show of one; the row's title carries the term.
const short = (value) =>
  typeof value === "string" && value.includes("/") ? value.split(/[/#]/).pop() : num(value);

const list = (values) => (values && values.length ? values.map(short).join(", ") : dash);

// A statement about the comparison itself, not about either side: shown before any number,
// because a reader who does not know two runs took different paths will read the durations.
function notes(lines) {
  const body = document.createElement("ul");
  body.className = "compare-notes";
  lines.forEach((line) => {
    const item = document.createElement("li");
    item.textContent = line;
    body.append(item);
  });
  return section("Not fully comparable", "read the deltas below with these in mind", body);
}

function renderModel(model) {
  const parts = document.createElement("div");
  parts.className = "views";
  if (model.values.length) {
    parts.append(section("Values", "a gain, a band, a limit", table(
      ["term", "property", "left", "right"],
      model.values.map((row) => ({
        cells: [row.name, row.property, num(row.left), num(row.right)],
        title: row.iri,
      })),
    )));
  }
  ["added", "removed"].forEach((kind) => {
    if (!model[kind].length) return;
    parts.append(section(`Terms ${kind}`, "", table(
      ["term", "kind"],
      model[kind].map((row) => ({ cells: [row.name, row.types.join(", ") || dash], title: row.iri })),
    )));
  });
  if (model.structure.length) {
    parts.append(section("Structure", "an operand, a controller set, a solver's limits", table(
      ["term", "property", "was", "now"],
      model.structure.map((row) => ({
        cells: [row.name, row.property, list(row.removed), list(row.added)],
        title: row.iri,
      })),
    )));
  }
  if (!parts.children.length) {
    const same = document.createElement("p");
    same.className = "view-note";
    same.textContent = "the two design graphs are identical";
    parts.append(same);
  }
  return section("Model", "over the graph, not the text", parts);
}

function renderCoordination(coordination) {
  const parts = document.createElement("div");
  parts.className = "views";
  parts.append(section("Activities", "a re-entry is its own row, never summed", table(
    ["activity", "left", "right", "delta"],
    coordination.activities.map((row) => ({
      cells: [row.name, seconds(row.left_s), seconds(row.right_s), signed(row.delta_s)],
      tint: row.left_s === null || row.right_s === null,
    })),
  )));
  if (coordination.gates.length) {
    parts.append(section("Gates", "waits and re-arms, never a verdict", table(
      ["gate", "left wait", "right wait", "delta", "re-arms"],
      coordination.gates.map((row) => ({
        cells: [
          row.name,
          seconds(row.left_waited_s),
          seconds(row.right_waited_s),
          signed(row.delta_s),
          `${row.left_rearms ?? dash} → ${row.right_rearms ?? dash}`,
        ],
        tint: row.left_fired !== row.right_fired,
      })),
    )));
  }
  const note = coordination.order_differs
    ? "the control-flow order differs — durations along different paths are not comparable"
    : "same control-flow order";
  return section("Coordination", note, parts);
}

function renderSignals(signals) {
  const parts = document.createElement("div");
  parts.className = "views";
  signals.activities.forEach((activity) => {
    const rows = [
      ...activity.signals.map((row) => ({
        cells: [
          row.signal,
          `peak ${num(row.left_peak)} → ${num(row.right_peak)}`,
          signed(row.delta_peak),
          `pk-pk ${signed(row.delta_pk_pk)} · rms ${signed(row.delta_rms)}`,
          row.gate ?? dash,
        ],
      })),
      ...activity.contact.map((row) => ({
        cells: [
          "contact",
          `peak ${num(row.left_peak_n)} N → ${num(row.right_peak_n)} N`,
          signed(row.delta_peak_n),
          `ripple ${num(row.left_ripple)} → ${num(row.right_ripple)}`,
          row.gate ?? dash,
        ],
      })),
      ...activity.modes.map((row) => ({
        cells: [
          `coherent mode · ${row.solver}`,
          `${num(row.left_mode_hz)} Hz → ${num(row.right_mode_hz)} Hz`,
          signed(row.delta_mode_hz),
          `amplitude ${num(row.left_mean_amplitude)} → ${num(row.right_mean_amplitude)}`,
          dash,
        ],
        // A mode that appeared or vanished is the finding; a mode that moved a band is not.
        tint: (row.left_mode_hz === null) !== (row.right_mode_hz === null),
      })),
    ];
    const entry = activity.entry ? ` (entry ${activity.entry + 1})` : "";
    parts.append(section(activity.name + entry, "", table(
      ["what", "left → right", "delta", "and", "gate"],
      rows,
    )));
  });
  if (signals.saturation.length) {
    parts.append(section("Saturation", "per run, as the report gives it", table(
      ["joint", "left ticks", "right ticks", "delta", "states"],
      signals.saturation.map((row) => ({
        cells: [
          `${row.solver} joint ${row.joint}`,
          num(row.left_ticks),
          num(row.right_ticks),
          signed(row.delta_ticks),
          [...new Set([...row.left_states, ...row.right_states])].join(", ") || dash,
        ],
      })),
    )));
  }
  return section("Signals", `restricted to ${signals.restricted_to}`, parts);
}

function render(slot, payload) {
  const parts = [];
  if (payload.comparable.notes.length) parts.push(notes(payload.comparable.notes));
  parts.push(renderModel(payload.model));
  if (payload.coordination.activities.length) parts.push(renderCoordination(payload.coordination));
  if (payload.signals.activities.length || payload.signals.saturation.length) {
    parts.push(renderSignals(payload.signals));
  }
  slot.replaceChildren(...parts);
}

// The whole comparison in the URL, so it can be pasted to a colleague: the left side is the
// run already open, the right side is what was picked here.
function remember(generation, run) {
  const view = new URLSearchParams(location.hash.slice(1));
  view.set("panel", "compare");
  for (const [key, value] of [["cmp_gen", generation], ["cmp_run", run]]) {
    if (value) view.set(key, value);
    else view.delete(key);
  }
  history.replaceState(null, "", `#${view}`);
}

async function load(shell, generation, run) {
  const slot = shell.querySelector(".compare-slot");
  if (!generation) return slot.replaceChildren();
  slot.textContent = "comparing…";
  // Runs are a pair or neither: a model diff against a generation with no run picked is
  // still worth showing, and the route says so in its notes.
  const query = new URLSearchParams({ left: state.generationPath, right: generation });
  if (run) {
    query.set("left_run", state.runPath);
    query.set("right_run", run);
  }
  render(slot, await api(`/api/compare?${query}`));
}

async function fillRuns(shell, generation, chosen) {
  const picker = shell.querySelector(".compare-run");
  picker.replaceChildren(new Option("model only", ""));
  if (!generation) return;
  const runs = await api(`/api/runs?path=${encodeURIComponent(generation)}`);
  runs.forEach((run) => picker.append(new Option(run.id, run.path)));
  if (chosen) picker.value = chosen;
}

export async function loadCompare() {
  const shell = $("#panel-compare");
  const view = new URLSearchParams(location.hash.slice(1));
  const wanted = `${state.runPath}|${view.get("cmp_gen") ?? ""}|${view.get("cmp_run") ?? ""}`;
  if (shell.dataset.loaded === wanted) return;
  shell.dataset.loaded = wanted;
  shell.querySelector(".compare-left").textContent = `${state.runPath.split("/").pop()} against`;
  const generations = await api("/api/generations");
  const genPicker = shell.querySelector(".compare-gen");
  genPicker.replaceChildren(new Option("compare with…", ""));
  generations.forEach((generation) =>
    genPicker.append(new Option(`${generation.name} · ${generation.path.split("/").pop()}`, generation.path)));
  genPicker.value = view.get("cmp_gen") ?? "";
  await fillRuns(shell, genPicker.value, view.get("cmp_run"));
  const runPicker = shell.querySelector(".compare-run");
  const show = async () => {
    remember(genPicker.value, runPicker.value);
    shell.dataset.loaded = `${state.runPath}|${genPicker.value}|${runPicker.value}`;
    await load(shell, genPicker.value, runPicker.value);
  };
  genPicker.onchange = async () => {
    await fillRuns(shell, genPicker.value, null);
    await show();
  };
  runPicker.onchange = show;
  if (genPicker.value) await load(shell, genPicker.value, runPicker.value);
}
